app/controllers/application_controller.rb

```diff
@@ -57,18 +57,25 @@ def user_fully_authenticated?
     user_signed_in? && current_user.two_factor_enabled? && is_fully_authenticated?
   end
 
-  def confirm_two_factor_setup
+  def confirm_two_factor_authenticated
+    authenticate_user!
+
     user_decorator = UserDecorator.new(current_user)
-    return if user_decorator.may_bypass_two_factor_setup?(session)
 
+    return if user_decorator.may_bypass_2fa?(session) || user_fully_authenticated?
+
+    return prompt_to_set_up_2fa unless current_user.two_factor_enabled?
+
+    prompt_to_enter_otp
```

Comment:
I know its probably because I'm new to ruby but I find this kind of logic hard to read.   This would make more sense to my non-ruby eyes:

```
def confirm_two_factor_authenticated                                     
  authenticate_user!      
  user_decorator = UserDecorator.new(current_user)
  if !user_fully_authenticated? && !user_decorator.may_bypass_2fa?(session)
    if current_user.two_factor_enabled?                                                   
      return prompt_to_enter_otp                                                          
    else                                                                                  
      return prompt_to_set_up_2fa                                                         
    end                                                                                   
  end  
  
  # falls of the end if everything is ok  
end
```